subsys/testsuite/ztest/include/ztest_test_new.h

```diff
@@ -0,0 +1,323 @@
+ */
+static inline void unit_test_noop(void)
+{
+}
+
+#define Z_ZTEST(suite, fn, t_options)                                                              \
+	static void _##suite##_##fn##_wrapper(void *data);                                         \
+	static void suite##_##fn(void);                                                            \
+	static STRUCT_SECTION_ITERABLE(ztest_unit_test, z_ztest_unit_test_##suite##_##fn) = {      \
+		.test_suite_name = STRINGIFY(suite),                                               \
+		.name = STRINGIFY(fn),                                                             \
+		.test = (_##suite##_##fn##_wrapper),                                               \
+		.thread_options = t_options,                                                       \
+	};                                                                                         \
+	static void _##suite##_##fn##_wrapper(void *data)                                          \
+	{                                                                                          \
+		ARG_UNUSED(data);                                                                  \
+		suite##_##fn();                                                                    \
+	}                                                                                          \
+	static inline void suite##_##fn(void)
+
+#define Z_ZTEST_F(suite, fn, t_options)                                                            \
+	static void _##suite##_##fn##_wrapper(void *data);                                         \
+	static void suite##_##fn(struct suite##_fixture *this);                                    \
+	static STRUCT_SECTION_ITERABLE(ztest_unit_test, z_ztest_unit_test_##suite##_##fn) = {      \
+		.test_suite_name = STRINGIFY(suite),                                               \
+		.name = STRINGIFY(fn),                                                             \
+		.test = (_##suite##_##fn##_wrapper),                                               \
+		.thread_options = t_options,                                                       \
+	};                                                                                         \
+	static void _##suite##_##fn##_wrapper(void *data)                                          \
+	{                                                                                          \
+		suite##_##fn((struct suite##_fixture *)data);                                      \
+	}                                                                                          \
+	static inline void suite##_##fn(struct suite##_fixture *this)
+
+#define ZTEST(suite, fn) Z_ZTEST(suite, fn, 0)
+
+#ifdef CONFIG_USERSPACE
+#define ZTEST_USER(suite, fn) Z_ZTEST(suite, fn, K_USER)
+#else
+#define ZTEST_USER(suite, fn)                                                                      \
+	BUILD_ASSERT(false, "Can't create userspace test without CONFIG_USERSPACE being enabled")
```

Comment:
imo this should fallback to ZTEST instead of assert. We have suites which have multiple configurations (with and without userspace) and `ztest_user_unit_test` falls to `ztest_unit_test`.

Reply:
K, was on the fence about this one. Done

Reply:
build assert still there.

Reply:
Sorry, forgot to push.